### ISSUE TITLE
Product Registration Doctrine ORM Implementation

### ISSUE THREAD

**Issue by MatthewVita** [openemr/openemr#430]
This picks up from the pending-merge `MatthewVita:doctrine_and_other_moderization_work` branch.

**Comment by MatthewVita** [comment]
@bradymiller I've never actually noticed this feature. It seems like it would be useful for folks at the front desk, right?

I may improve the UX a bit if you don't mind.

**Comment by MatthewVita** [comment]
@juggernautsei now that we're rolling with creating the entities/repositories... do you know of another good table to model with doctrine? That way, we both can work on a table at the same time and help each other out.

**Comment by juggernautsei** [comment]
I am not sure what is meant by "good table". The first table was the version table because it was simple. So if we are keeping with that mind set a simple table prices ( has three primary keys) and has only 4 columns.
Groups has just one primary key and 3 columns. Product registration has three columns as well with one primary key.

Apologies for having to show ignorance here.  I am still trying to play catch up to what you are doing. As Brady put it that there is a high learning curve. This modeling reminds me of when I use to work with code igniter. I am excited to do more work with OOP.

I have reviewed the notes in the version model.  So in creating a new model, I would start with 

      namespace TABLE_NAME;

Is this correct assumption?

Also, we are just going to be doing the modeling, right? I remember in building code igniter projects I would go through and build all the models before building the code that used the models if there was an existing database. Other models can be added as needed but the initial phase like now to build all the models possible. 

I can set out to model the groups table and see what I can come up with (darn that means I have to stop putting off that doctrine course I marked a month ago to get through).

Tally HO!

**Comment by juggernautsei** [comment]
I am going to lean on your guidance. I started on the product registration model. Here is what I have so far.

        namespace repositories;

       use entities\Product_registration;
       use Doctrine\ORM\EntityRepository;

       class ProductRegistrationRepository extends EntityRepository{
           /**
            * Creates the sole entry that the product has been registered
            * The table has one unique key
            * The table should have only one entry
            * @param $registered
            * 
            */
         public function create(Product_registration $registered){
         	$this->_em->presist($registered);
         	$this->_em->flush(); //tells doctrine to write info to table
        	return true;
              } 
          }

**Comment by juggernautsei** [comment]
The use entities should be the database table name right?

**Comment by MatthewVita** [comment]
> I am not sure what is meant by "good table"

I should have been more clear. What I mean is we started with a very simple table that was only used in a few places and we can up the difficulty as we go along (try to model entities that will make use of JOINs, TRANSACTIONs, custom Doctrine query language, etc).

> Apologies for having to show ignorance here. I am still trying to play catch up to what you are doing. As Brady put it that there is a high learning curve. This modeling reminds me of when I use to work with code igniter. I am excited to do more work with OOP.

Code Ignitor was/is awesome. Very simply and moved a lot of PHP codebases in the right direction! Take your time getting this stuff down... there is a high learning curve but it will be so easy over time.

> I have reviewed the notes in the version model. So in creating a new model, I would start with namespace TABLE_NAME;

Just put it in `entities/table_name` and `repositories/table_name_repository.php`.


> Also, we are just going to be doing the modeling, right? I remember in building code igniter projects I would go through and build all the models before building the code that used the models if there was an existing database. Other models can be added as needed but the initial phase like now to build all the models possible.

This approach is know as "Database First". There are two ways to incorporate ORMs "Database First" where the database/tables are already set up and "Code First" where the models themselves create the database/tables. Both have their merits, but you'll find that "Database First" is what you'll use with existing codebases. Therefore, we simply model the existing schema.

> product registration model.

This is a great entity for you to model. Be sure to return the status of if `create` was successful or not, however.

> autoload_classmap.php

Don't edit this directly. This is generated by Composer. Please read Brady's excellent guide here: http://www.open-emr.org/wiki/index.php/Composer

> doesTableExist

This method is not needed. We needed this for version because of the requirements for upgrades. Not needed with product registration.

> The use entities should be the database table name right?

Yup.

**Comment by MatthewVita** [comment]
> entities\Product_registration

Rename to `entities\ProductRegistration` for consistency.

**Comment by MatthewVita** [comment]
BTW please make sure you are branched off of the moderization branch that I made (yes there is a type in the name "modernization").

**Comment by MatthewVita** [comment]
Also, before I introduced the `openemr/services` directory, I was put the service for the product registration in `openemr/interface/product_registration/product_registration_service.php`. If you feel like it, you can move it to `openemr/service/product_registration_service.php` to encourage MVC in our codebase.

If you have any troubles with the production registration feature implementation, you can blame me because I was the one that implemented it. I'm on the hook :).

**Comment by MatthewVita** [comment]
I renamed this ticket from "ONotes Doctrine ORM Implementation" to "Product Registration Doctrine ORM Implementation". This ticket is yours now... I'm going to create another ticket called "ONotes ORM Implementation" that I'll work on in parallel.

Sorry for the confusion. #ProjectManagement

**Comment by juggernautsei** [comment]
No problem. Thanks!

**Comment by juggernautsei** [comment]
entities\Product_registration
Rename to entities\ProductRegistration for consistency.

There is a issue that needs to be addressed because if the entities is the database table name and the table name has underscores in it. What do we do?

**Comment by juggernautsei** [comment]
autoload_classmap.php
Don't edit this directly. This is generated by Composer. Please read Brady's excellent guide here: http://www.open-emr.org/wiki/index.php/Composer

I went to the page and I do have composer installed on my server.
However, I am concerned that the original install of composer was not for OpenEMR directory. It was for a different directory so now if I run the command that Brady suggest. Will it do the right thing?

@bradymiller

**Comment by MatthewVita** [comment]
> There is a issue that needs to be addressed because if the entities is the database table name and the table name has underscores in it. What do we do?

Hey. This should be resolved by using `@Table(name="table_name")` on the entity itself. See the following: https://github.com/MatthewVita/openemr/blob/a2c2eccd85583fa153cb613d4ed84254fbd399a5/entities/version.php#L31

Thanks,
Matthew

**Comment by MatthewVita** [comment]
> However, I am concerned that the original install of composer was not for OpenEMR directory. It was for a different directory so now if I run the command that Brady suggest. Will it do the right thing?

I don't understand. Do mean you simply ran `composer install` but perhaps in a subdirectory such as `openemr/interface`?

**Comment by bradymiller** [comment]
@juggernautsei ,

In theory, you shouldn't need to run composer since those directories/namespaces are already set up to pick up PSR4 compliant classes.

However, to answer your question, as long as you run composer from the openemr main directory(where the composer.json file is), you will be fine.

Since you are just updating the autoloader, just need to do the 1 command in the autoloader section on the wiki page (again, though, this shouldn't be necessary).

**Comment by juggernautsei** [comment]
@bradymiller 
You confirmed what I was thinking as long as composer is ran from the location that contains the script need to be used. It would work. 

I 'm not far enough along in understanding to say that the auto loader would pickup any additional files that were added to the repository folder.

**Comment by juggernautsei** [comment]
@MatthewVita 
I am really good at mimicking other code that I see and reshaping it. But I am going to ask  you for a little understanding along the way. 

You wrote:
Hey. This should be resolved by using @Table(name="table_name") on the entity itself. See the following: 

I looked at the code and does that mean that doctrine reads the comments section for instructions?

**Comment by juggernautsei** [comment]
Question:
I should have created the entities file first?

        namespace entities;

			use Doctrine\ORM\Mapping\Entity;
			use Doctrine\ORM\Mapping\Table;
			use Doctrine\ORM\Mapping\Column;
			use Doctrine\ORM\Mapping\Id;


			/**
			* @Table(name="product_registration")
			* @Entity(repositoryClass="repositories\ProductRegistrationRepository")
			*/

			class ProductRegistration {
			/**
			 * Default constructor.
			 */
			public function __construct(){

			}

			/**
			 * @Column(name="p_registration_id"), type="char", length=36 nullable=false, options={"default" : 0})
			 */
			 private $registration_id;

			/**
			 * @Column(name="p_email"), type="varchar", length=255 nullable=false, options={"default" : 0})
			 */
			 private $email;

			/**
			 * @Column(name="p_opt_out"), type="tinyint", length=1 nullable=false, options={"default" : 0})
			 */
			 private $opt_out

			/**
			 * Getter for registration_id.
			 *
			 * return registration_id number
			 */
			public function getRegistration_id() {
			    return $this->registration_id;
			}

			/**
			 * Setter for registration_id.
			 *
			 * @param registration_id string
			 */
			public function setRegistration_id($value) {
			    $this->registration_id = $value;
			}     

			/**
			 * Getter for email.
			 *
			 * return email string
			 */
			public function getEmail() {
			    return $this->email;
			}

			/**
			 * Setter for email.
			 *
			 * @param email string
			 */
			public function setEmail($value) {
			    $this->email = $value;
			}     

			/**
			 * Getter for opt_out.
			 *
			 * return opt_out string
			 */
			public function getOpt_out() {
			    return $this->opt_out;
			}

			/**
			 * Setter for opt_out.
			 *
			 * @param opt_out number
			 */
			public function setOpt_out($value) {
			    $this->opt_out = $value;
			}

			/**
			 * ToString of the entire object.
			 *
			 * @return object as string
			 */
			public function __toString() {
			    return "registration_id: '" . $this->getRegistration_id() . "' " .
			           "email: '" . $this->getEmail() . "' " .
			           "opt_out" . $this->getOpt_out() . "' " ;
			}
		}

**Comment by juggernautsei** [comment]
The current product_registration_services.php file should not be moved to the services folder because it is not of the same class that we are currently build because it uses the old system right?

**Comment by MatthewVita** [comment]
Glad you are mimicking the code and learning quickly!

> I looked at the code and does that mean that doctrine reads the comments section for instructions?

Yup!

>  private $opt_out

Missing a semicolon.

> getRegistration_id

Just use camelCase to stay consistent.

> The current product_registration_services.php file should not be moved to the services folder because it is not of the same class that we are currently build because it uses the old system right?

We'll have to swap out the old sql statements with the new repository you're making, but it can totally be moved to the services folder. Autoloading it will now work. For example in openemr/interface/product_registration/,
you'll replace

```
require_once($GLOBALS['fileroot'] . "/interface/product_registration/product_registration_service.php");

...
...

class ProductRegistrationController {

    private $productRegistrationService;

    public function __construct() {
        $this->productRegistrationService = new ProductRegistrationService();
        ...
```

with:
```

...
...

class ProductRegistrationController {

    private $productRegistrationService;

    public function __construct() {
        $this->productRegistrationService = new \services\ProductRegistrationService();
        ...
```

**Comment by juggernautsei** [comment]
Made corrections to the product_registration.php in the entities folder.

/**
	 * @Table(name="product_registration")
	 * @Entity(repositoryClass="repositories\ProductRegistrationRepository")
	 */

	class ProductRegistration {
		/**
		 * Default constructor.
		 */
		public function __construct(){

		}

		/**
		 * @Column(name="p_registration_id"), type="char", length=36 nullable=false, options={"default" : 0})
		 */
		 private $registration_id;

		/**
		 * @Column(name="p_email"), type="varchar", length=255 nullable=false, options={"default" : 0})
		 */
		 private $email;

		/**
		 * @Column(name="p_opt_out"), type="tinyint", length=1 nullable=false, options={"default" : 0})
		 */
	     private $opt_out;

	    /**
	     * Getter for registration_id.
	     *
	     * return registration_id number
	     */
	    public function getRegistrationid() {
	        return $this->registration_id;
	    }

	    /**
	     * Setter for registration_id.
	     *
	     * @param registration_id string
	     */
	    public function setRegistrationid($value) {
	        $this->registration_id = $value;
	    }     

	    /**
	     * Getter for email.
	     *
	     * return email string
	     */
	    public function getEmail() {
	        return $this->email;
	    }

	    /**
	     * Setter for email.
	     *
	     * @param email string
	     */
	    public function setEmail($value) {
	        $this->email = $value;
	    }     

	    /**
	     * Getter for opt_out.
	     *
	     * return opt_out string
	     */
	    public function getOptout() {
	        return $this->opt_out;
	    }

	    /**
	     * Setter for opt_out.
	     *
	     * @param opt_out number
	     */
	    public function setOptout($value) {
	        $this->opt_out = $value;
	    }

	    /**
	     * ToString of the entire object.
	     *
	     * @return object as string
	     */
	    public function __toString() {
	        return "registration_id: '" . $this->getRegistrationid() . "' " .
	               "email: '" . $this->getEmail() . "' " .
	               "opt_out" . $this->getOptout() . "' " ;
	    }
	}

**Comment by juggernautsei** [comment]
Would you please share with me how this works in doctrine?

/**
* @Table(name="product_registration")
* @Entity(repositoryClass="repositories\ProductRegistrationRepository")
*/

I am assuming that the @Table tells doctrine to read this command and sets the name of the table to what ever comes after the = sign. 

How does that play out in the over all project?

I don't want to guess at the @Entity. Could you just point me to documentation or give a brief?

Thanks for all you help!

**Comment by MatthewVita** [comment]
> @Table(name="product_registration")

Your understanding is correct with the `@Table`.

> How does that play out in the over all project?

It doesn't affect the overall project. Only when the `repository` is referencing the entity does the `@Table` `name` value gets referenced. Put simply, `repository` and `entity manager` map the the plain PHP entity object and the table together (whether that is with SELECTs or INSERTs/UPDATEs).

> give a brief?

`@Entity`s must be linked to a `repository`. Basically, an `@Entity` without a `repository` is just a boring old PHP object. Only when it is linked with a `repository` and `entity manager` will it become "database aware", so to speak.

**Comment by juggernautsei** [comment]
Thank you so much for that.

**Comment by juggernautsei** [comment]
To make sure that I am on the right trail. The three main folders that we are focused on are the entities, services and repository. The entities folder is where the database table description lives. The repository is where the work is done that send the information database via the flush command.

**Comment by juggernautsei** [comment]
> We'll have to swap out the old sql statements with the new repository you're making, but it can totally be moved to the services folder. Autoloading it will now work. For example in openemr/interface/product_registration/, you'll replace

Should the controller be moved with the services file to the services folder?

**Comment by MatthewVita** [comment]
> Thank you so much for that.

Anytime.

> The entities folder is where the database table description lives.

Yup and where `JOIN`s are called out. I will have an example of this with the `onotes` entity I'm working on. I will be working on it tomorrow.

> The repository is where the work is done that send the information database via the flush command.

Yup. `EntityManager::flush` causes write operations against the database to be executed. It is the last piece to be called to make `persist`, `merge`, `delete`, etc work.

Services are obviously for business logic.

> Should the controller be moved with the services file to the services folder?

Controllers will eventually go into the `controllers/` folder, but that's a longer story (long story short, we will deal with that after doctrine project is done). Services should move by themselves for now.

Thanks,
Matthew

**Comment by juggernautsei** [comment]
Matthew, 
not sure what to do now. The files that are created are the three main folders. How would I go about testing the code to see if it works. Or should I post it in a git for your review of what I have done so far?

**Comment by MatthewVita** [comment]
Please post it on a branch for me to see. Thanks!

**Comment by juggernautsei** [comment]
[Here is the registration code](https://github.com/openemr/openemr/compare/master...juggernautsei:product_registration_model?expand=1)

**Comment by MatthewVita** [comment]
Hi @juggernautsei,

Sorry I've been quiet. I've been focused on https://github.com/openemr/openemr/pull/443

I'm going to look at your branch now!

Thanks,
Matthew

**Comment by MatthewVita** [comment]
@juggernautsei,

> Again, I am using the versions_service.php file as my guide. I noticed in line 78 there is an update function to update the version entry. Is $version and object?

Yup!

Here's some notes/questions:

- Why are there `p_`s in front the column names in the entity? For example `p_registration_id`.
- You'll need a fetch method in the repository because of the `public function getProductStatus() { $row = sqlQuery('SELECT * FROM product_registration LIMIT 1');`
- Did you try out your create method in the places where the INSERTs are done in the ProductRegistrationService ?
- I noticed you are using $snake_case instead of $camelCase in some areas. Which style should we use? I personally $camelCase but I see that both styles are used in OpenEMR.

Thanks,
Matthew

**Comment by juggernautsei** [comment]
> I noticed you are using $snake_case instead of $camelCase in some areas

I am trying to be consistant in using the camel case but it will take awhile to work the snake case out of my writings. 

> Why are there p_s in front the column names in the entity? For example p_registration_id
I was copying what you have in the version entity v_patch, v_minor. So the same question to you. Why are there v_m in the entity file?

**Comment by MatthewVita** [comment]
> I was copying what you have in the version entity v_patch, v_minor. So the same question to you. Why are there v_m in the entity file?

The version table has been around for quiet some time and whoever introduced it prefixed each column with `v_`. Have no idea. You don't need to do that with your entity.

> No I have not. I am not confident enough to try it.
Give it a try. Let me know if you get stuck anywhere.


BTW, I'm back to working on the onotes and it will be the first JOIN via Doctrine! I just modeled out the `onote` entity and am now working on JOINing it with `user`! Have to model out `user` entity first, of course. I will update you on my progress in the https://github.com/openemr/openemr/issues/432 ticket, of course.

Thanks,
Matthew Vita

**Comment by juggernautsei** [comment]
See that is how tightly I was following the example code that I did not look to see if the table had the v_m in them. Now that I look I can see how silly what I did may have looked when you first saw it.

**Comment by MatthewVita** [comment]
It happens!

**Comment by MatthewVita** [comment]
check out these few lines of code... hooking up our first JOIN https://github.com/openemr/openemr/pull/453/files#diff-9e45f07c55aaf4970d22bfe4b5a1171bR67

**Comment by juggernautsei** [comment]
It looks like the direction I was heading when doing the research on joins.

**Comment by MatthewVita** [comment]
@juggernautsei,

How's this going?

Thanks,
Matthew

**Comment by juggernautsei** [comment]
Not well. I am getting a little spead thin and I don't really know where I am going with the doctrine transformation. I am trying to keep up but you are going faster then I can get around to working with (not your fault or complaining). I have some time today to get back to this but like I asked before what do I do next now that the repository/entity/service are built?

**Comment by MatthewVita** [comment]
Hi @juggernautsei,

Maybe we can do a Google Hangouts tomorrow at 6pm UTC -5 to work on this together. If that won't work, please try to work on the following:

Let's start out with just the autoloading and `SELECT` pieces. Open up `openemr/interface/product_registration/product_registration_service.php` and add the following field and update the constructor:

```php
private $repository;

public function __construct() {
  $database = \common\database\Connector::Instance();
  $entityManager = $database->entityManager;
  $this->repository = $entityManager->getRepository('\entities\ProductRegistration');
}
```

Now you can edit the following:

```php
public function getProductStatus() {
  //$row = sqlQuery('SELECT * FROM product_registration LIMIT 1');
  $productRegistration = $repository->findFirst();
  ...
  ...
```

`findFirst` is actually not provided by EntityRepository, so add the following method to your repository:

```php
public function findFirst() {
  $results = $this->_em->getRepository($this->_entityName)->findAll();
    if (!empty($results)) {
      return $results[0];
    }

    return null;
}
```

...once this is working, let's do the INSERT piece next.

Thanks,
Matthew

**Comment by juggernautsei** [comment]
I'll make myself available at 6pm UTC to work on this together and in the mean time I will follow the steps you have listed.

**Comment by juggernautsei** [comment]
Just to be talking out loud to help my understanding, the constructor is where the database connection lives. The findFirst function is created to deliver record 0. 
I have these pieces in place.

**Comment by MatthewVita** [comment]
@juggernautsei just sent you a calendar/hangout invite for our 6pm UTC -5 meeting

**Comment by MatthewVita** [comment]
It includes the video link

**Comment by MatthewVita** [comment]
https://hangouts.google.com/call/cnzjctguvfa4xnfj5onk5hp5qie

**Comment by MatthewVita** [comment]
@juggernautsei this (https://github.com/openemr/openemr/pull/471) should be the complete implementation to compare your branch against for the product registration doctrine implementation. I went ahead and put it all together so you can compare your work to it along the way.

Disregard the calendar template stuff that got checked in... oops!
